src/engine/routing_algorithms/alternative_path_mld.cpp

```diff
@@ -0,0 +1,936 @@
+
+    // Store all alternative packed paths (if there are any).
+    auto into = std::back_inserter(weighted_packed_paths);
+    std::transform(begin(candidate_vias) + 1, last_filtered, into, extract_packed_path_from_heaps);
+
+    // Filter packed paths with heuristics
+
+    auto alternative_paths_last = end(weighted_packed_paths);
+
+    alternative_paths_last =
+        filterPackedPathsByLocalOptimalityHeuristic(weighted_packed_paths[0],
+                                                    forward_heap, // paths for s, via
+                                                    reverse_heap, // paths for via, t
+                                                    begin(weighted_packed_paths) + 1,
+                                                    alternative_paths_last);
+
+    const auto number_of_alternative_paths =
+        alternative_paths_last - (begin(weighted_packed_paths) + 1);
+
+    // Todo: refactor - this is ugly af
+
+    auto number_of_filtered_alternative_paths =
+        std::min(static_cast<std::size_t>(max_number_of_alternatives_to_unpack),
+                 static_cast<std::size_t>(number_of_alternative_paths));
+
+    for (std::size_t i = 1; i < number_of_filtered_alternative_paths; ++i)
+    {
+        for (std::size_t j = 0; j < i; ++j)
+        {
+            alternative_paths_last =
+                filterPackedPathsByCellSharingHeuristic(weighted_packed_paths[j],
+                                                        partition,
+                                                        begin(weighted_packed_paths) + i,
+                                                        alternative_paths_last);
+        }
+    }
+
+    number_of_filtered_alternative_paths = std::min(
+        static_cast<std::size_t>(max_number_of_alternatives_to_unpack),
+        static_cast<std::size_t>(alternative_paths_last - (begin(weighted_packed_paths) + 1)));
+
+    // ^ refactor
+
+    const auto paths_first = begin(weighted_packed_paths);
+    const auto paths_last = begin(weighted_packed_paths) + 1 + number_of_filtered_alternative_paths;
+    const auto number_of_packed_paths = paths_last - paths_first;
+
+    // Todo: benchmark how many packed paths we can unnpack in our time budget.
+
+    std::vector<WeightedViaNodeUnpackedPath> unpacked_paths;
+    unpacked_paths.reserve(number_of_packed_paths);
+
+    // Note: re-uses (read: destroys) heaps; we don't need them from here on anyway.
+    unpackPackedPaths(paths_first,
+                      paths_last,
+                      std::back_inserter(unpacked_paths),
+                      search_engine_data,
+                      facade,
+                      phantom_node_pair);
+
+    //
+    // Filter and rank a second time. This time instead of being fast and doing
+    // heuristics on the packed path only we now have the detailed unpacked path.
+    //
+
+    auto unpacked_paths_last = end(unpacked_paths);
```

Comment:
This should be refactored into an own function that works on a range of unpacked paths.